docs/reference/files/metadata-yaml-file.rst

```diff
@@ -0,0 +1,896 @@
+
+.. code-block:: yaml
+
+    # (Optional) A string (or a list of strings) containing a link (or links) to the charm source code.
+    source: <url> | {ref}`<urls>]
+
+
+``storage``
+-----------
+
+.. code-block:: yaml
+
+    # (Optional) Storage requests for the charm
+    storage:
+      # Each key represents the name of the storage
+      <storage name>:
+
+          # (Required) Type of the requested storage
+          type: filesystem | block
+
+          # (Optional) Description of the storage requested
+          description: <description>
+
+          # (Optional) The mount location for filesystem stores. For multi-stores
+          # the location acts as the parent directory for each mounted store.
+          location: <location>
+
+          # (Optional) Indicates if all units of the application share the storage
+          shared: true | false
+
+          # (Optional) Indicates if the storage should be made read-only (where possible)
+          read-only: true | false
+
+          # (Optional) The number of storage instances to be requested
+          multiple:
+              range: <n> | <n>-<m> | <n>- | <n>+
+
+          # (Optional) Minimum size of requested storage in forms G, GiB, GB. Size
+          # multipliers are M, G, T, P, E, Z or Y. With no multiplier supplied, M
+          # is implied.
+          minimum-size: <n>| <n><multiplier>
+
+          # (Optional) List of properties, only supported value is "transient"
+          properties:
+              - transient
+
+
+``subordinate``
+---------------
+
+.. code-block:: yaml
+
+    # (Optional) True if the charm is meant to be deployed as a subordinate to a
+    # principal charm
+    subordinate: true | false
+
+
+``summary``
+-----------
+
+.. code-block:: yaml
+
+    # (Required) A short, one-line description of the charm
+    summary: <summary>
+
+
+``terms``
+---------
+
+.. code-block:: yaml
+
+    # (Optional) A list of terms that any charm user must agree with
+    terms:
+        - <term>
+
+
+``website``
+-----------
+
+.. important::
+
+    In :ref:`recipes <charmcraft-yaml-file>`, this is now the ``links.website`` key.
+
+.. code-block:: yaml
+
+    # (Optional) A string (or a list of strings) containing a link (or links) to project websites.
+    # In general this is likely to be the upstream project website, or the formal website for the
+    # charmed offering.
+    website: <url> | [<urls>]
+
+
+``Other keys``
+--------------
+
+In addition to the keys mentioned above, a ``metadata.yaml`` file may also contain other
+arbitrary keys. These can serve to keep track of other choices a charmer might make. In
+some cases these become semi-official, being adopted by many charmers and even
+incorporated into CI processes. An example is
+:literalref:`upstream source <https://github.com/canonical/charmcraft/blob/b22fcdba3b894004468abfbf45caa54d93fbf7d0/charmcraft/templates/init-simple/metadata.yaml.j2#L40-L43>`_.
```

Comment:
```suggestion
incorporated into CI processes.
```

Let's cut this, because it isn't appropriate to provide such a specific example for an open-ended syntax.

Reply:
This was added on purpose as an example of how thisv unofficial key is used by convention to point to this upstream source. 

Links to specific pieces of existing charm yaml and code from the charm ecosystem are a pattern we've deliberately tried to use in both Charmcraft and Ops docs as they are a lightweight yet very powerful way to make a point.

Reply:
We were wondering about this one. The linked code was removed from the repo a while back, so it looked to me like it was a historical artifact that's no longer current.

The main reason for removal was the link breaks the linter. There's no way around that at the moment, and GH doesn't have a built-in shortener.

Since the code isn't live in the repo and therefore the reader can't be reassured of its currency, I suggest we simply copy it verbatim here.

Reply:
As I was saying, linking to real examples of charm code has been a pattern we've followed in other places in the docs too, on purpose and for a good reason -- it makes it possible to keep the docs quite generic and clean while still giving people a sense of what the content covered in the docs looks like in the wild. The fact that the linter doesn't like it doesn't seem to me a good reason to undo this pattern.

Reply:
So that we don't lose this important snippet, can we copy the code in verbatim? We do that quite a bit in Starcraft docs.

Reply:
Since we're now using sphinx I wonder if we can use a `literalinclude`  to the path? I don't know if that works with a path outside the docs root, but maybe a symlink could work around that?

I would prefer to include other text on the page rather than linking to it, but if we can avoid having another copy to maintain that would also be preferable.

Reply:
> if we can use a literalinclude to the path

I think we're conflating intersphinx features and literalinclude. The latter expects a local file, so we'd have to first get the text from a specific Git commit – not simple.

JJ's latest commit is a simple way to address the question for now: [f45b51b](https://github.com/canonical/charmcraft/pull/2083/commits/f45b51b672c0b7ee9da1e5795cf72a5193a1d91b)